csrc/multidevice/utils.cpp

```diff
@@ -136,42 +127,178 @@ int64_t numDeviceDims(const TensorView* tv) {
+      [](Val* val) { return val->as<IterDomain>(); });
+  return inputs_as_iter_domains;
+}
+
+bool overlaps(
+    const std::vector<IterDomain*>& a,
+    const std::unordered_set<IterDomain*>& b) {
+  return std::any_of(
+      a.begin(), a.end(), [&](IterDomain* id) { return b.count(id); });
+}
+
+} // namespace
+
 bool haveDifferentShardings(
     const TensorView* producer,
-    const TensorView* consumer) {
+    const TensorView* consumer,
+    const IdModel& id_model) {
   // cpu scalars are not required to have a mesh
   if (producer->isCpuScalar() || consumer->isCpuScalar()) {
     return false;
   }
+
   // exit early in the unsharded case for performance
   if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
     return false;
   }
+
   // If device mesh are different, the Expr is resharding
-  if (!(producer->getDeviceMesh() == consumer->getDeviceMesh())) {
+  if (producer->getDeviceMesh() != consumer->getDeviceMesh()) {
     return true;
   }
-  // Create a map between producer's and consumer's IterDomains. We iterate
-  // over producer's iterdomain and compare sharding type with consumer's
-  // iterdomain
+
+  // The rest of this function tries to do the following: for each pair of
+  // logical-domain-mapped IterDomains (i.e. those mapped by
+  // PairwiseLogicalDomainMap), check if they are sharded consistently. If not,
+  // returns true. For example,
+  //
+  //   a: iDIDx{M}, iK
+  //   b: iK, iDIDy{N}
+  //   c = matmul(a, b): iDIDx{M}, iDIDy{N}
+  //
+  // haveDifferentShardings(a, c) only cares about iM, which is
+  // logical-domain-mapped, but not iK or iN, which are not
+  // logical-domain-mapped.
+  //
+  // One challenge is that DID parallelization doesn't always
+  // happen on the root/logical IterDomains. For example, a root/logical
+  // IterDomain may be outer-split by the number of devices, and only the outer
+  // split gets parallelized on DID.
+  //
+  //   logical: iM
+  //   loop: iDIDx{D}, iM/D
+  //
+  // Therefore, we collect all the loop IterDomains that depend on the
+  // logical-domain-mapped IterDomains, and check if they are DID-parallelized
+  // consistently.
   const std::unordered_map<IterDomain*, IterDomain*>& p2c =
       PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
-  for (auto p_id : producer->getLogicalDomain()) {
-    const auto i = p2c.find(p_id);
+  std::unordered_set<IterDomain*> mapped_p_logical_ids;
+  mapped_p_logical_ids.reserve(p2c.size());
+  std::unordered_set<IterDomain*> mapped_c_root_ids;
+  mapped_c_root_ids.reserve(p2c.size());
+  for (IterDomain* p_logical_id : producer->getLogicalDomain()) {
+    const auto i = p2c.find(p_logical_id);
     if (i == p2c.end()) {
-      // This happens e.g. when `p_id` is squeezed or is a product of a
-      // reduction. Even if `p_id` is parallelized on DID, the dimension is
-      // size-1 and doesn't trigger resharding.
+      // This happens e.g. when `p_logical_id` is squeezed. Even if
```

Comment:
This also happens with a reduction domain. If the producer is an output of a reduction, its logical domain has a reduction iter domain, which is not inherit to the consumer tensor. Not sure if it would matter, though.

Reply:
Correct. I accidentally copied an old comment -- I didn't mean to exclude reduction.